### PR TITLE
Adding bootstrapping for MSVSMON

### PR DESCRIPTION
**Pull request by alanch-ms**

### CODE REVIEW

File: Python/Product/Uwp/Project/PythonUwpProjectConfig.cs [hunk 1]
```diff
@@ -711,6 +684,79 @@ class PythonUwpProjectConfig :
...
+            IVsAppContainerBootstrapper4 bootstrapper = (IVsAppContainerBootstrapper4)ServiceProvider.GlobalProvider.GetService(typeof(SVsAppContainerProjectDeploy));
+            return (IVsAppContainerBootstrapperResult)bootstrapper.BootstrapForDebuggingAsync(this.GetProjectUniqueName(), targetDevice, BootstrapMode.UniversalBootstrapMode, this.GetRecipeFile(), logger: null).GetResult();
+        }
+
+        private string GetProjectUniqueName() {
+            EnvDTE.DTE dte = (EnvDTE.DTE)Package.GetGlobalService(typeof(EnvDTE.DTE));
+            EnvDTE.SolutionBuild sb = dte.Solution.SolutionBuild;
+            string projectUniqueName = string.Empty;
+            foreach (String s in (Array)sb.StartupProjects) {
+                // There should only be one startup project so get the first one we find.
```
Comment (int19h) [inline Python/Product/Uwp/Project/PythonUwpProjectConfig.cs]: That's not true - you can set multiple startup projects in VS (right-click on solution node -> "Set Startup Projects").

File: Python/Product/Uwp/Project/PythonUwpProjectConfig.cs [hunk 2]
```diff
@@ -711,6 +684,79 @@ class PythonUwpProjectConfig :
...
+        private IVsAppContainerBootstrapperResult BootstrapForDebuggingSync(string targetDevice) {
+            IVsAppContainerBootstrapper4 bootstrapper = (IVsAppContainerBootstrapper4)ServiceProvider.GlobalProvider.GetService(typeof(SVsAppContainerProjectDeploy));
+            return (IVsAppContainerBootstrapperResult)bootstrapper.BootstrapForDebuggingAsync(this.GetProjectUniqueName(), targetDevice, BootstrapMode.UniversalBootstrapMode, this.GetRecipeFile(), logger: null).GetResult();
+        }
+
+        private string GetProjectUniqueName() {
+            EnvDTE.DTE dte = (EnvDTE.DTE)Package.GetGlobalService(typeof(EnvDTE.DTE));
+            EnvDTE.SolutionBuild sb = dte.Solution.SolutionBuild;
+            string projectUniqueName = string.Empty;
+            foreach (String s in (Array)sb.StartupProjects) {
```
Comment (int19h) [inline Python/Product/Uwp/Project/PythonUwpProjectConfig.cs]: Might as well cast to `string[]` then.

File: Python/Product/Uwp/Project/PythonUwpProjectConfig.cs [hunk 3]
```diff
@@ -711,6 +684,79 @@ class PythonUwpProjectConfig :
...
+                // There should only be one startup project so get the first one we find.
+                projectUniqueName = s;
+                break;
+            }
+            if (projectUniqueName.Equals(string.Empty)) {
```
Comment (int19h) [inline Python/Product/Uwp/Project/PythonUwpProjectConfig.cs]: `==` is strictly better here (shorter, and will handle null should it ever occur).

File: Python/Product/Uwp/Project/PythonUwpProjectConfig.cs [hunk 4]
```diff
@@ -711,6 +684,79 @@ class PythonUwpProjectConfig :
...
+            }
+
+            return recipeFile;
+        }
+
+        private string GetStringPropertyValue(string propertyName) {
+            IVsSolution solution = (IVsSolution)Package.GetGlobalService(typeof(SVsSolution));
+            IVsHierarchy hierarchy;
+            solution.GetProjectOfUniqueName(GetProjectUniqueName(), out hierarchy);
+            IVsBuildPropertyStorage bps = hierarchy as IVsBuildPropertyStorage;
```
Comment (int19h) [inline Python/Product/Uwp/Project/PythonUwpProjectConfig.cs]: Should be a `(T)` cast if it's expected to never fail.

File: Python/Product/Uwp/Project/PythonUwpProjectConfig.cs [hunk 5]
```diff
@@ -920,29 +979,48 @@ class PythonUwpProjectConfig :
...
-            this.outputWindow = null;
-
-            return VSConstants.S_OK;
+            outputWindow = null;
         }
 
-        private void NotifyEndDeploy(int success) {
-            foreach (IVsDeployStatusCallback callback in deployCallbackCollection) {
-                callback.OnEndDeploy(success);
+        private IEnumerable<IVsDeployStatusCallback> GetSinkCollection() {
```
Comment (int19h) [inline Python/Product/Uwp/Project/PythonUwpProjectConfig.cs]: This whole thing looks like `Enumerable.ToList` under a `lock`.

**Comment by alanch-ms** [comment]
Thanks for the feedback.  I should have address all the feedback in the 2nd commit.  Should I merge the two commits together?

**Comment by zooba** [comment]
No, separate commits are fine.

File: Python/Product/Uwp/Project/PythonUwpProjectConfig.cs [hunk 6]
```diff
@@ -920,31 +979,43 @@ class PythonUwpProjectConfig :
...
-            return VSConstants.S_OK;
+            return true;
         }
 
         private void NotifyEndDeploy(int success) {
-            foreach (IVsDeployStatusCallback callback in deployCallbackCollection) {
-                callback.OnEndDeploy(success);
+            try {
+                foreach (IVsDeployStatusCallback callback in GetSinkCollection()) {
+                    callback.OnEndDeploy(success);
```
Comment (zooba) [inline Python/Product/Uwp/Project/PythonUwpProjectConfig.cs]: Why could they be null above but not here?

File: Python/Product/Uwp/Project/PythonUwpProjectConfig.cs [hunk 7]
```diff
@@ -920,31 +979,43 @@ class PythonUwpProjectConfig :
...
+                    this.deployOp = null;
+                }
             }
 
             outputWindow = null;
         }
 
+        private IEnumerable<IVsDeployStatusCallback> GetSinkCollection() {
+            lock (syncObject) {
+                return this.deployCallbackCollection.Cast<IVsDeployStatusCallback>().ToList();
```
Comment (zooba) [inline Python/Product/Uwp/Project/PythonUwpProjectConfig.cs]: Using `OfType<T>` instead of `Cast<T>` will remove nulls from the resulting sequence.

**Comment by alanch-ms** [comment]
Address more code review feedback.  Thanks!

**Comment by int19h** [comment]
:+1: (but wait until Steve okays it, too)

**Pull request merged**